numba/cuda/kernels/transpose.py

```diff
@@ -0,0 +1,53 @@
+from numba import cuda
+from numba.cuda.cudadrv.driver import driver
+from numba import numpy_support as nps
+import math
+
+def transpose(a, b=None):
+    """Compute the transpose of 'a' and store it into 'b', if given,
+    and return it. If 'b' is not given, allocate a new (host) array
+    and return that.
+
+    This implements the algorithm documented in 
+    http://devblogs.nvidia.com/parallelforall/efficient-matrix-transpose-cuda-cc/"""
+
+    if not b:
+        cols, rows = a.shape
```

Comment:
'a' isn't necessarily rectangular (so the dimensions are reversed). This implies that the strides change, too.